tnylpo-convert.1
.\"
.\" Copyright (c) 2019 Georg Brein. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright notice,
.\"    this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" 3. Neither the name of the copyright holder nor the names of its
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH tnylpo-convert 1 2024-10-12
.SH NAME
tnylpo-convert \- converts text files to and from the CP/M format
.SH SYNOPSIS
.HP
.B tnylpo-convert 
.RB [ -aeinz ]
.RB [ -f
.IR <config-file> ]
[
.RB ( -u
.RI ( <fn>
|
.BR - )
|
.B -c
.IR <fn> )
[
.RB ( -u
.RI ( <fn>
|
.BR - )
|
.B -c
.IR <fn> )
] ]
.HP
.B tnylpo-convert -h
.SH DESCRIPTION
tnylpo-convert is a companion program of
.BR tnylpo (1)
and allows the conversion of text files from the format used by the
host operating system to the format used by CP/M applications running
under
.BR tnylpo (1)
(resp.
.IR "vice versa" ).
.PP
The difference between the Unix text file format and CP/M text files is
threefold:
.IP \(em
Unix uses a single
.B <lf>
(0x0a) character as line end marker, while CP/M uses the sequence
.B <cr> <lf>
(0x0d 0x0a).
.IP \(em
CP/M uses the first
.B <sub>
(0x1a) character in the file as end of file marker, except if the logical end
of the text file coincides with a 128 byte record boundary
.IP \(em
CP/M uses 8-bit (or 7-bit) characters, while most recent
operating systems use wide or multibyte character sets.
.PP
tnylpo-convert shares its configuration files with
.BR tnylpo (1),
but uses only the options for the character set definitions
.RB ([ alt ]
.BR char ,
.RB [ alt ]
.BR charset ,
and
.BR unprintable ),
on which it bases its character translation. Both the configuration file
syntax and the way tnylpo-convert finds its configuration are described
in the man page of
.BR tnylpo (1).
.SH OPTIONS
.TP
.B -a
use the alternate character set from the configuration file
for character translation
.TP
.BI -c " <cpm-textfile>"
use the text file
.I <cpm-textfile>
in CP/M format as source or target of the conversion
.TP
.B -e
treat unconvertible characters in the source file as an error
.TP
.BI -f " <config-file>"
tell tnylpo-convert explicitly which configuration file to use
.TP
.B -h
ask tnylpo-convert to show a short command line synopsis
.RB ( -h
cannot be used in combination with any other command line option)
.TP
.B -i
ignore all unconvertible characters by silently dropping them
.TP
.B -n
do not pad CP/M text files with
.B <sub>
(0x1a) characters; CP/M programs running under
.BR tnylpo (1)
will still accept such files, since the BDOS emulation automatically pads
files not ending on a record boundary with
.B <sub>
(0x1a) bytes on input
.TP
.BI -u " <textfile>"
use the text file
.I <textfile>
in host system format as source or target of the conversion; if
.B -
is given instead of
.IR <textfile> ,
.B stdin
is read resp.
.B stdout
is written
.TP
.B -z
always append a
.B <sub>
(0x1a) character as an end of file marker to CP/M text files, even if
they end on a record boundary
.PP
The options
.B -e
and
.B -i
are mutually exclusive; if neither is given, tnylpo-convert will drop
unconvertible characters, but issue a warning. The options
.B -z
and
.B -n
can be combined: If both
.B -z
and
.B -n
are given, the resulting file will be terminated with exactly one
.B <sub>
(0x1a) byte.
.B
.PP
tnylpo-convert takes no positional arguments. The
.B -c
and 
.B -u
options may be given at most twice; the first occurrence defines the
source file, the second occurrence the target file of the conversion.
If the target file or both source and target files are missing,
.B -u -
is used instead. The same file name may be given both as source and
as target, since tnylpo-convert uses a temporary file for output, which
is renamed to the target file name on successful completion (this is not
done if
.B stdout
is used for output).
.SH EXIT STATUS
tnylpo-convert exits with status 0 if it didn't encounter
command line, configuration, or I/O errors; otherwise (or if
unconvertible characters are encountered and the
.B -e
option is specified on the command line), status 1 is returned.
.SH FILES
.SS ./.tnylpo.conf
.SS ~/.tnylpo.conf
default configuration files, see
.BR tnylpo (1)
.SH EXAMPLES
.B tnylpo-convert -f myconf.conf -e -c hugo.txt -u -
.PP
converts the contents of the CP/M text file
.B hugo.txt
to the format and character set of the host operating system and
prints them on
.BR stdout .
Character set information is taken from the configuration file
.BR myconf.conf ,
and any unconvertible characters in
.B hugo.txt
will cause the conversion to fail.
.PP
.B tnylpo-convert -u test.pas -c test.pas -z
.PP
converts the Unix text file
.B test.pas
to CP/M format; the default configuration file (if found) is
used for character set information. The resulting file is always
terminated with at least one
.B <sub>
(0x1a) character.
.SH AUTHOR
Georg Brein
.RB ( tnylpo@gmx.at )
.SH SEE ALSO
.BR tnylpo (1)


tnylpo-convert.c
/*
 * Copyright (c) 2019 Georg Brein. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */


#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <stdarg.h>
#include <string.h>
#include <locale.h>
#include <wchar.h>
#include <errno.h>

#include <unistd.h>

#include "tnylpo.h"


/*
 * always append ^Z when writing a CP/M output file, even if the
 * data ends on a 128 byte CP/M record boundary
 */
static int append_cntrlz = 0;
/*
 * omit padding a CP/M output file with ^Z bytes
 */
static int omit_padding = 0;
/*
 * how to handle unconvertible characters
 */
static int ignore_convert = 0;
static int error_convert = 0;
/*
 * source and target file names, types, and file pointers
 */
static char *source_name = NULL, *target_name = NULL;
static int source_unix = 0, target_unix = 0;
static int source_stdin = 0, target_stdout = 0;
static FILE *source_fp = NULL, *target_fp = NULL;
/*
 * number of bytes written to a target file in CP/M format
 */
static off_t target_size = 0;


/*
 * program name for error messages
 */
static const char *prog_name = NULL;


/*
 * write a message to stderr
 */
void
perr(const char *format, ...) {
	va_list params;
	va_start(params, format);
	fprintf(stderr, "%s: ", prog_name);
	vfprintf(stderr, format, params);
	fprintf(stderr, "\n");
	va_end(params);
}


/*
 * display a short usage summary
 */
void
usage(void) {
	perr("usage: %s [ <options> ] [ <source> [ <target> ] ]",
	    prog_name);
	perr("valid <options> are");
	perr("    -a              use alternate charset");
	perr("    -e              treat unconvertible characters as error");
	perr("    -f <fn>         read configuration file");
	perr("    -i              ignore unconvertible characters");
	perr("    -n              do not pad CP/M files with ^Z bytes");
	perr("    -z              always terminate CP/M files with ^Z");
	perr("<source> or <target> are");
	perr("    -u ( <fn> | - ) text file in host OS format");
	perr("    -c <fn>         text file in CP/M format");
	perr("if <source> resp. <target> is omitted, \"-u -\"  is assumed");
}


/*
 * print error message if a command line option has been defined more than once
 */
static void
only_once(int c) {
	perr("option -%c may be specified only once", c);
}


/*
 * get the configuration
 *
 * After parsing the command line, further configuration values are
 * read from the optional configuration file
 */
static int
get_config(int argc, char **argv) {
	int rc = 0, opt;
	char *cfn = NULL;
	opterr = 0;
	while ((opt = getopt(argc, argv, "ac:ef:inu:z")) != EOF) {
		switch (opt) {
		case 'a':
			/*
			 * use alternate character set
			 */
			if (charset) {
				only_once('a');
				rc = (-1);
			} else {
				charset = 1;
			}
			break;
		case 'c':
		case 'u':
			if (source_name && target_name) {
				perr("source and target already specified");
				rc = (-1);
			}
			if (! strcmp(optarg, "-")) {
				if (opt != 'u') {
					perr("stdin/stdout only allowed for "
					    "Unix files");
					rc = (-1);
				} else {
					if (! source_name) {
						source_name = "<stdin>";
						source_stdin = 1;
						source_unix = 1;
					} else {
						target_name = "<stdout>";
						target_stdout = 1;
						target_unix = 1;
					}
				}
			} else {
				if (! source_name) {
					source_name = optarg;
					source_stdin = 0;
					source_unix = (opt == 'u');
				} else {
					target_name = optarg;
					target_stdout = 0;
					target_unix = (opt == 'u');
				}
			}
			break;
		case 'e':
			/*
			 * treat unconvertible characters as error
			 */
			if (error_convert) {
				only_once('e');
				rc = (-1);
			} else {
				error_convert = 1;
			}
			break;
		case 'f':
			/*
			 * configuration file
			 */
			if (cfn) {
				only_once('f');
				rc = (-1);
			}
			cfn = optarg;
			break;
		case 'i':
			/*
			 * ignore unconvertible characters
			 */
			if (ignore_convert) {
				only_once('i');
				rc = (-1);
			} else {
				ignore_convert = 1;
			}
			break;
		case 'n':
			/*
			 * omit padding a CP/M target file with ^Z bytes
			 */
			if (omit_padding) {
				only_once('n');
				rc = (-1);
			} else {
				omit_padding = 1;
			}
			break;
		case 'z':
			/*
			 * always append ^Z to the end of a CP/M target file
			 */
			if (append_cntrlz) {
				only_once('z');
				rc = (-1);
			} else {
				append_cntrlz = 1;
			}
			break;
		case '?':
			perr("invalid option -%c", optopt);
			rc = (-1);
			break;
		}
	}
	/*
	 * defaults for source and target are Unix format files read from
	 * stdin resp. written to stdout
	 */
	if (! source_name) {
		source_name = "<stdin>";
		source_unix = 1;
		source_stdin = 1;
	}
	if (! target_name) {
		target_name = "<stdout>";
		target_unix = 1;
		target_stdout = 1;
	}
	/*
	 * positional arguments are not allowed on the command line
	 */
	if (argc != optind) {
		perr("positional arguments are not allowed");
		rc = (-1);
	}
	/*
	 * -i and -e are incompatible
	 */
	if (ignore_convert && error_convert) {
		perr("options -i and -e are mutually exclusive");
		rc = (-1);
	}
	/*
	 * command line error: print usage information and die
	 */
	if (rc) {
		usage();
		goto premature_exit;
	}
	/*
	 * read the optional configuration file
	 */
	rc = read_config(cfn);
premature_exit:
	return rc;
}


/*
 * write a character to the CP/M file; keeps track of the output file size
 */
static void
write_cpm(unsigned char uc) {
	target_size += fwrite(&uc, sizeof uc, 1, target_fp);
}


/*
 * For once, no comment.
 */
int
main(int argc, char **argv) {
	int rc = 0, t, last_was_cr, convert_error = 0;
	wint_t wc;
	unsigned char uc;
	char *temp_name = NULL;
	prog_name = base_name(argv[0]);
	if (! setlocale(LC_CTYPE, "")) {
		perr("setlocale(LC_CTYPE) failed");
		rc = (-1);
		goto premature_exit;
	}
	/*
	 * if the only parameter is -h, print the usage summary and exit
	 */
	if (argc == 2 && ! strcmp(argv[1], "-h")) {
		usage();
		goto premature_exit;
	}
	/*
	 * parse command line and read configuration file
	 */
	if (get_config(argc, argv)) {
		perr("command line or configuration error");
		goto premature_exit;
	}
	/*
	 * open source file
	 */
	if (source_stdin) {
		source_fp = stdin;
	} else {
		source_fp = fopen(source_name, source_unix ? "r" : "rb");
		if (! source_fp) {
			perr("couldn't open %s: %s", source_name,
			    strerror(errno));
			rc = (-1);
			goto premature_exit;
		}
	}
	/*
	 * open target file
	 */
	if (target_stdout) {
		target_fp = stdout;
	} else {
		/*
		 * build temporary target file name
		 */
		temp_name = alloc(strlen(target_name) + 30);
		sprintf(temp_name, "%s.temp.%lu", target_name,
		    (unsigned long) getpid());
		target_fp = fopen(temp_name, target_unix ? "w" : "wb");
		if (! target_fp) {
			perr("couldn\'t open %s: %s", temp_name,
			    strerror(errno));
			rc = (-1);
			goto premature_exit;
		}
	}
	/*
	 * main loop: copy characters
	 */
	if (source_unix && target_unix) {
		/*
		 * Unix to Unix: no translation whatsoever
		 */
		while ((wc = fgetwc(source_fp)) != WEOF) fputwc(wc, target_fp);
	} else if (source_unix && ! target_unix) {
		/*
		 * Unix to CP/M: convert characters, translate LF to CR/LF
		 */
		while ((wc = fgetwc(source_fp)) != WEOF) {
			if (wc == L'\n') {
				write_cpm(0x0d /* CR */);
				write_cpm(0x0a /* LF */);
			} else {
				t = to_cpm(wc);
				if (t == (-1)) {
					convert_error = 1;
				} else {
					write_cpm((unsigned char) t);
				}
			}
		}
	} else if (! source_unix && target_unix) {
		/*
		 * CP/M to Unix: stop at first SUB, convert characters,
		 * translate CR/LF to LF
		 */
		last_was_cr = 0;
		while (fread(&uc, sizeof uc, 1, source_fp) &&
		     uc != 0x1a /* SUB */) {
			if (uc == 0x0a /* LF */) {
				/*
				 * bare LF and CR/LF are translated to LF
				 */
		     		fputwc(L'\n', target_fp);
				last_was_cr = 0;
			} else {
				/*
				 * let bare CRs survive
				 */
				if (last_was_cr) fputwc(L'\r', target_fp);
				if (uc == 0x0d /* CR */) {
					/*
					 * remember CR
					 */
					last_was_cr = 1;
				} else {
					/*
					 * convert other character
					 */
					wc = from_cpm(uc);
					if (wc == WEOF) {
						convert_error = 1;
					} else {
						fputwc(wc, target_fp);
					}
					last_was_cr = 0;
				}
			}
		}
		/*
		 * copy last bare CR
		 */
		if (last_was_cr) fputwc(L'\r', target_fp);
	} else {
		/*
		 * CP/M to CP/M: stop at first SUB
		 */
		while (fread(&uc, sizeof uc, 1, source_fp) &&
		     uc != 0x1a /* SUB */) write_cpm(uc);
	}
	/*
	 * terminate CP/M format output
	 */
	if (! target_unix) {
		if (append_cntrlz) write_cpm(0x1a /* SUB */);
		if (! omit_padding) {
			while (target_size % 128) write_cpm(0x1a /* SUB */);
		}
	}
	/*
	 * check for source read errors
	 */
	if (ferror(source_fp)) {
		perr("read error on %s: %s", source_name, strerror(errno));
		rc = (-1);
	}
	/*
	 * check for target write errors
	 */
	if (ferror(target_fp)) {
		perr("write error on %s: %s", temp_name ? temp_name :
		    target_name, strerror(errno));
		rc = (-1);
	}
	/*
	 * handle unconvertible characters
	 */
	if (convert_error) {
		if (! ignore_convert) {
			perr("%s%s contains untranslateable characters",
			    error_convert ? "" : "warning: ", source_name);
			if (error_convert) rc = (-1);
		}
	}
premature_exit:
	if (source_fp && ! source_stdin) {
		/*
		 * close the source file, if it isn't stdin
		 */
		if (fclose(source_fp)) {
			perr("couldn\'t close %s: %s", source_name,
			    strerror(errno));
			rc = (-1);
		}
	}
	if (target_fp && ! target_stdout) {
		/*
		 * close the target file, if it isn't stdout
		 */
		if (fclose(target_fp)) {
			perr("couldn\'t close %s: %s", temp_name,
			    strerror(errno));
			rc = (-1);
		}
		if (! rc) {
			/*
			 * rename the temporary file to the target name
			 */
			if (rename(temp_name, target_name)) {
				perr("couldn\'t rename %s to %s: %s",
				    temp_name, target_name, strerror(errno));
				rc = (-1);
			}
		}
		/*
		 * in case of error, try to remove the temporary file
		 */
		if (rc) remove(temp_name);
	}
	free(temp_name);
	exit(rc ? EXIT_FAILURE : EXIT_SUCCESS);
	return 0;
}
